plugins/inputs/radius/radius.go

```diff
@@ -0,0 +1,134 @@
+package radius
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type Radius struct {
+	Servers         []string        `toml:"servers"`
+	Username        config.Secret   `toml:"username"`
+	Password        config.Secret   `toml:"password"`
+	Secret          config.Secret   `toml:"secret"`
+	ResponseTimeout config.Duration `toml:"response_timeout"`
+	Log             telegraf.Logger `toml:"-"`
+	client          radius.Client
+}
+
+//go:embed sample.conf
+var sampleConfig string
+
+func (r *Radius) SampleConfig() string {
+	return sampleConfig
+}
+
+func (r *Radius) Init() error {
+	if len(r.Servers) == 0 {
+		r.Servers = []string{"127.0.0.1:1812"}
+	}
+
+	r.client = radius.Client{
+		Retry: 0,
+	}
+
+	return nil
+}
+
+func (r *Radius) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	for _, server := range r.Servers {
+		wg.Add(1)
+		go func(server string) {
+			defer wg.Done()
+			acc.AddError(r.pollServer(acc, server))
+		}(server)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (r *Radius) pollServer(acc telegraf.Accumulator, server string) error {
+	// Create the fields for this metric
+	host, port, err := net.SplitHostPort(server)
+	if err != nil {
+		return fmt.Errorf("splitting host and port failed: %w", err)
+	}
+	tags := map[string]string{"source": host, "source_port": port}
+	fields := make(map[string]interface{})
+
+	secret, err := r.Secret.Get()
+	if err != nil {
+		return fmt.Errorf("getting secret failed: %w", err)
+	}
+	defer config.ReleaseSecret(secret)
+
+	username, err := r.Username.Get()
+	if err != nil {
+		return fmt.Errorf("getting username failed: %w", err)
+	}
+	defer config.ReleaseSecret(username)
+
+	password, err := r.Password.Get()
+	if err != nil {
+		return fmt.Errorf("getting password failed: %w", err)
+	}
+	defer config.ReleaseSecret(password)
+
+	// Create the radius packet with PAP authentication
+	packet := radius.New(radius.CodeAccessRequest, secret)
+	err = rfc2865.UserName_Set(packet, username)
+	if err != nil {
+		return fmt.Errorf("setting username for radius auth failed: %w", err)
+	}
+	err = rfc2865.UserPassword_Set(packet, password)
+	if err != nil {
+		return fmt.Errorf("setting password for radius auth failed: %w", err)
+	}
+
+	// Do the radius request
+	ctx := context.Background()
+	if r.ResponseTimeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(r.ResponseTimeout))
+		defer cancel()
+	}
+
+	startTime := time.Now()
+	response, err := r.client.Exchange(ctx, packet, server)
+	duration := time.Since(startTime)
+
+	if err != nil {
+		r.Log.Warnf("error on new request to %s : %s", server, err)
+		fields["responsetime_ms"] = time.Duration(r.ResponseTimeout).Milliseconds()
+		tags["response_code"] = ""
```

Comment:
How about passing the error on to the user and not clutter Telegraf's log, i.e.
```suggestion
	if err != nil {
		fields["responsetime_ms"] = time.Duration(r.ResponseTimeout).Milliseconds()
		if errors.Is(err, context.DeadlineExceeded) {
			tags["response_code"] = "timeout"
		} else {
			tags["response_code"] = "error"
			fields["error"] = err.Error()
		}
```

Reply:
Or alternatively

```suggestion

	if err != nil {
		if !errors.Is(err, context.DeadlineExceeded) {
			return err
		}
		fields["responsetime_ms"] = time.Duration(r.ResponseTimeout).Milliseconds()
		tags["response_code"] = "timeout"
```

Reply:
I dont have any strong opinion on this, I guess best solution is what is usually used in these cases in telegraf, which Im not that familiar with.

The alternative one looks cleaner (and no additional field).

Only possible issue would be that some users might expect the response_code to be only from the set of possible response codes from the radius protocol standard (but the old solution with empty string there was already against that too).

I can only speak for our use case, the errors/warnings in logs could as well may have been proper errors as you suggest, we only used them to initially setup/debug the plugin - they are useful when user is first setting up the plugin and possibly when something breaks (likely on remote server side - password changed, radius user deleted, radius server changed IP etc).

One thing good about not returning an error but returning standard response with the timeout duration at least for our current use case is that when ANY error occurs in the monitoring setup, and the user is monitoring only the response time, he still gets an alert, without having to have any other monitoring in place to also check the telegraf errors etc. But monitoring for missing data points should cover that... so nothing major.

I guess Ill leave this one to the telegraf team, so its more inline with the standard conventions you use.

Maybe it would be good to modify the plugin description to clarify that the plugin is meant to monitor SUCCESSFUL logins only instead of "The Radius plugin collects radius authentication response times."  Because right now even on wrong password you get the timeout time, even though the response time in strict sense was lower, but without Access-Accept code.

Our global tacacs/radius servers use many different backends for auth (global AD servers, local auth etc), so only successfull auth response time is interesting since that covers the whole timespan of the request as it goes through all the dependent systems and various networks/regions to see real auth response times.  Non-existent user that gets immediately declined but with proper response code that is not Access-Accept is not interesting to monitor. For some it might be though. I guess lets leave that for possible future feature requests? 

BTW, did not check deeply, but is the testcontainer issue resolution compatible with use cases where app listens on both TCP and UDP on the same port ? Could cause some clashes where TCP mapped port is overwritten with UDP in the c.Ports[port] construct, or vice versa, but did not test the actual behavior.

Reply:
@powersj any thoughts?

Reply:
Going with the alternative option - I think having a tag a user could filter on to avoid the non-timeout conditions gives users a way to avoid those situations. And non-timeouts should error.